Add validator package

torives opened the pull request:
Depends on #444 and #560. Closes #553.

torives committed:
feat: add validator package

torives committed:
feat(validator): add service to the supervisor

torives committed:
feat(validator): run as a standalone binary

torives committed:
refactor(repository): update test name

torives committed:
fix(startup): properly wrap error

fmoura commented on internal/validator/validator.go:
```diff
@@ -0,0 +1,302 @@
+type ValidatorRepository interface {
+	// GetAllRunningApplications returns a slice with the applications currently
+	// being validated by the node.
+	GetAllRunningApplications(ctx context.Context) ([]Application, error)
```
The way repository interface is partitioned makes this method be declared more than once. This could lead to some inconsistencies. 

But as a repository redesign will take place in the newer future we can prevent this pointing out a single doc source for some methods

fmoura commented on internal/validator/validator.go:
```diff
@@ -0,0 +1,302 @@
+		return fmt.Errorf("failed to get running applications. %w", err)
+	}
+
+	for idx := range apps {
```
Any strong reason to use `for idx := range apps` instead of `for _,app := range apps` ?

torives replied:
To avoid creating an extra copy of the struct.

fmoura commented on internal/validator/validator.go:
```diff
@@ -0,0 +1,302 @@
+			// this is the first epoch, return the pristine claim
+			claim, _, err := merkle.CreateProofs(nil, MAX_OUTPUT_TREE_HEIGHT)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to create proofs. %w", err)
```
Add `app address` and `epoch index` to message, to make clear for which epoch this originates from

torives replied:
Done in [986d2c5](https://github.com/cartesi/rollups-node/pull/554/commits/986d2c5ae59b7067115b9fb27c49873a84ebd1b8)

torives commented on internal/validator/validator.go:
```diff
@@ -0,0 +1,302 @@
+			epoch,
+			outputs,
+		); err != nil {
+			return fmt.Errorf("failed to store claim and proofs. %w", err)
```
Done in [986d2c5](https://github.com/cartesi/rollups-node/pull/554/commits/986d2c5ae59b7067115b9fb27c49873a84ebd1b8)

torives commented on internal/validator/validator.go:
```diff
@@ -0,0 +1,302 @@
+
+		// ...and compare it to the hash calculated by the Validator
+		if *machineClaim != *claim {
+			return errors.New("validator claim does not match Cartesi Machine claim")
```
Done in [986d2c5](https://github.com/cartesi/rollups-node/pull/554/commits/986d2c5ae59b7067115b9fb27c49873a84ebd1b8)

torives commented on internal/validator/validator.go:
```diff
@@ -0,0 +1,302 @@
+			epoch.AppAddress,
+		)
+		if err != nil {
+			return fmt.Errorf("failed to get the machine claim. %w", err)
```
Done in [986d2c5](https://github.com/cartesi/rollups-node/pull/554/commits/986d2c5ae59b7067115b9fb27c49873a84ebd1b8)

torives committed:
fix(repository): correct field order

When reading data from the returned database row in the GetPreviousEpoch function, `claimHash` was being assigned to `transactionHash` and vice-versa.

renan061 commented on internal/validator/validator.go:
```diff
@@ -0,0 +1,317 @@
+	SetEpochClaimAndInsertProofsTransaction(
+		ctx context.Context,
+		epoch Epoch,
+		outputs []Output,
```
We should return slices of pointers: `[]*Output`.
We can do this later though.

torives replied:
I'll add it to the post-alpha to-do list 👍🏽

renan061 commented on internal/validator/validator.go:
```diff
@@ -0,0 +1,317 @@
+	GetProcessedEpochs(ctx context.Context, application Address) ([]Epoch, error)
+	// GetLastInputOutputHash returns the outputs Merkle tree hash calculated
+	// by the Cartesi Machine after it processed the last input in the epoch.
+	GetLastInputOutputHash(
```
`GetLastInputOutput*s*Hash`

torives replied:
Done.

renan061 commented on internal/validator/validator.go:
```diff
@@ -0,0 +1,317 @@
+		epoch.ClaimHash = claim
+
+		// store the epoch and proofs in the database
+		if err := v.repository.SetEpochClaimAndInsertProofsTransaction(
```
I prefer to no use this syntax if it means breaking multiple lines.
Do `err = ...` in the previous line, than do `if err != nil ...`.

torives replied:
Done

renan061 commented on internal/validator/validator.go:
```diff
@@ -0,0 +1,317 @@
+	return "validator"
+}
+
+func (s *ValidatorService) Start(ctx context.Context, ready chan<- struct{}) error {
```
We can start using the Poller. It is already merged.
Add a TODO somewhere if you don't do it in this PR.

torives replied:
implementing Poller now would require small adjustments in several files, so I will refrain from doing it here to keep this PR leaner. But this is already planned for post-alpha release, don't worry 👍🏽

renan061 reviewed (dismissed):
Some comments that you can address in another PR at your discretion.
